src/core/utils/flash.cpp

```diff
@@ -0,0 +1,266 @@
+    Record  record;
+
+    record.Init(aKey, aFirst);
+    record.SetData(aValue, aValueLength);
+
+    if ((mSwapSize - record.GetSize()) < mSwapUsed)
+    {
+        Swap();
+        VerifyOrExit((mSwapSize - record.GetSize()) >= mSwapUsed, error = OT_ERROR_NO_BUFS);
+    }
+
+    otPlatFlashWrite(&GetInstance(), mSwapIndex, mSwapUsed, &record, record.GetSize());
+
+    record.SetAddCompleteFlag();
+    otPlatFlashWrite(&GetInstance(), mSwapIndex, mSwapUsed, &record, sizeof(RecordHeader));
+
+    mSwapUsed += record.GetSize();
+
+exit:
+    return error;
+}
+
+bool Flash::DoesValidRecordExist(uint32_t aOffset, uint16_t aKey) const
+{
+    RecordHeader record;
+    bool         rval = false;
+
+    for (; aOffset < mSwapUsed; aOffset += record.GetSize())
+    {
+        otPlatFlashRead(&GetInstance(), mSwapIndex, aOffset, &record, sizeof(record));
+
+        if (record.IsValid() && record.IsFirst() && (record.GetKey() == aKey))
+        {
+            ExitNow(rval = true);
+        }
+    }
+
+exit:
+    return rval;
+}
+
+void Flash::Swap(void)
+{
+    uint8_t  dstIndex  = !mSwapIndex;
+    uint32_t dstOffset = kSwapMarkerSize;
+    Record   record;
+
+    otPlatFlashErase(&GetInstance(), dstIndex);
+
+    for (uint32_t srcOffset = kSwapMarkerSize; srcOffset < mSwapUsed; srcOffset += record.GetSize())
+    {
+        otPlatFlashRead(&GetInstance(), mSwapIndex, srcOffset, &record, sizeof(RecordHeader));
+
+        VerifyOrExit(record.IsAddBeginSet());
+
+        if (!record.IsValid() || DoesValidRecordExist(srcOffset + record.GetSize(), record.GetKey()))
+        {
+            continue;
+        }
+
+        otPlatFlashRead(&GetInstance(), mSwapIndex, srcOffset, &record, record.GetSize());
+        otPlatFlashWrite(&GetInstance(), dstIndex, dstOffset, &record, record.GetSize());
+        dstOffset += record.GetSize();
+    }
+
+exit:
+    otPlatFlashWrite(&GetInstance(), dstIndex, 0, &sSwapActive, sizeof(sSwapActive));
+    otPlatFlashWrite(&GetInstance(), mSwapIndex, 0, &sSwapInactive, sizeof(sSwapInactive));
+
+    mSwapIndex = dstIndex;
+    mSwapUsed  = dstOffset;
+}
+
+otError Flash::Delete(uint16_t aKey, int aIndex)
+{
+    otError      error = OT_ERROR_NOT_FOUND;
+    int          index = 0;
+    RecordHeader record;
+
+    for (uint32_t offset = kSwapMarkerSize; offset < mSwapUsed; offset += record.GetSize())
+    {
+        otPlatFlashRead(&GetInstance(), mSwapIndex, offset, &record, sizeof(record));
+
+        if ((record.GetKey() != aKey) || !record.IsValid())
+        {
+            continue;
+        }
+
+        if (record.IsFirst())
+        {
+            index = 0;
+        }
+
+        if ((aIndex == index) || (aIndex == -1))
+        {
+            record.SetDeleted();
+            otPlatFlashWrite(&GetInstance(), mSwapIndex, offset, &record, sizeof(record));
+            error = OT_ERROR_NONE;
+        }
+
+        if ((index == 1) && (aIndex == 0))
+        {
+            record.SetFirst();
```

Comment:
Is it necessary to mark an item as "first", shouldn't it being the first found indicate that it is the first?

Write and erase operations can be pretty expensive in time, power, and line wearing.

Reply:
The existing `Set()` implementation does not mark older records as deleted/invalid. Instead, `Set()` uses the "first" flag to indicate that all older records should be ignored. Does that make sense?